File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

bool isVisited[7];
bool isCheck[10000000];
int cnt = 0;

bool isPrime(string now)
{
    int num = stoi(now);
    
    if(num==0 || num==1) return false;
    for(int i=2; i<=sqrt(num); i++)
    {
        if(num%i==0) return false;
    }
    return true;
    
}
void dfs(string numbers, string now)
{
    if(now.size() > 0 && !isCheck[stoi(now)])
    {
        isCheck[stoi(now)] = true;
        if(isPrime(now)) cnt++;
    }
    for(int i=0; i<numbers.size(); i++)
    {
        if(isVisited[i]) continue;
        isVisited[i] = true;
        dfs(numbers, now+numbers[i]);
        isVisited[i] = false;
    }
}
int solution(string numbers) {
    int answer = 0;
    dfs(numbers, "");
    
    answer = cnt;
    return answer;
}

File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    public int solution(String numbers) {
        int answer = 0;
        HashSet <Integer> set = new HashSet <> ();
        permutation("", numbers, set);
        
        while(set.iterator().hasNext())
        {
            int a = set.iterator().next();
            set.remove(a);
            if(a==2) answer++;
            else if(a%2!=0 && isPrime(a)) answer++;
        }
        
        return answer;
    }
    
    public boolean isPrime(int n) {
        if(n==0 || n==1) return false;
        for(int i=3; i<=Math.sqrt(n); i+=2)
        {
            if(n%i==0) return false;
        }
        return true;
    }
    public void permutation(String prefix, String str, HashSet<Integer> set)
    {
        int n = str.length();
        
        if(prefix.length()>0)
        {
            set.add(Integer.parseInt(prefix));
        }
        
        for(int i=0; i<n; i++)
        {
            permutation(prefix + str.charAt(i), str.substring(0,i)+str.substring(i+1, n), set);
        }
    }
}